scipy/stats/_stats_py.py

```diff
@@ -9018,64 +9029,136 @@ def confidence_interval(self, confidence_level=0.95):
+            low = np.where(valid_index, np.take_along_axis(x, low_index), np.nan)
+            high = np.full(shape, np.inf)
         elif alternative == 'two-sided':
             p = (1 - confidence_level) / 2
-            low_index = int(bd.ppf(p)) - 1
-            low = x[low_index] if low_index >= 0 else np.nan
-            high_index = int(bd.isf(p))
-            high = x[high_index] if high_index < n else np.nan
-
+            low_index = (bd.ppf(p)).astype(int) - 1
+            valid_index = low_index >= 0
+            low_index[~valid_index] = 0
+            low = np.where(valid_index, np.take_along_axis(x, low_index), np.nan)
+            high_index = bd.isf(p).astype(int)
+            valid_index = high_index < n
+            high_index[~valid_index] = 0
+            high = np.where(valid_index, np.take_along_axis(x, high_index), np.nan)
+
+        args = self._axis, self._axis_none, self._keepdims, self._ndim, self._nan_out
+        low = _quantile_test_postprocess(low, *args)
+        high = _quantile_test_postprocess(high, *args)
         return ConfidenceInterval(low, high)
 
 
-def quantile_test_iv(x, q, p, alternative):
+def quantile_test_iv(x, q, p, alternative, axis, keepdims):
 
     x = np.atleast_1d(x)
-    message = '`x` must be a one-dimensional array of numbers.'
-    if x.ndim != 1 or not np.issubdtype(x.dtype, np.number):
+    message = '`x` must be an array of numbers.'
+    if not np.issubdtype(x.dtype, np.number):
         raise ValueError(message)
 
     q = np.array(q)[()]
-    message = "`q` must be a scalar."
-    if q.ndim != 0 or not np.issubdtype(q.dtype, np.number):
+    message = "`q` must be a scalar or array of numbers."
+    if not np.issubdtype(q.dtype, np.number):
         raise ValueError(message)
 
     p = np.array(p)[()]
-    message = "`p` must be a float strictly between 0 and 1."
-    if p.ndim != 0 or p >= 1 or p <= 0:
+    message = "`p` must be a scalar or array of floats."
+    if not np.issubdtype(p.dtype, np.inexact):
+        raise ValueError(message)
+
+    axis_none = axis is None
+    ndim = max(x.ndim, p.ndim)
+    if axis_none:
+        x = np.ravel(x)
+        q = np.ravel(q)
+        p = np.ravel(p)
+        axis = 0
+    elif np.iterable(axis) or int(axis) != axis:
+        message = "`axis` must be an integer or None."
         raise ValueError(message)
+    elif (axis >= ndim) or (axis < -ndim):
+        message = "`axis` is not compatible with the shapes of the inputs."
+        raise ValueError(message)
+    axis = int(axis)
 
     alternatives = {'two-sided', 'less', 'greater'}
     message = f"`alternative` must be one of {alternatives}"
     if alternative not in alternatives:
         raise ValueError(message)
 
-    return x, q, p, alternative
+    if keepdims not in {None, True, False}:
+        message = "If specified, `keepdims` must be True or False."
+        raise ValueError(message)
+
+    x = np.sort(x, axis=axis, stable=False)
```

Comment:
`stable` not in NumPy < 2.0. We can add it back when this is translated to array API.
```suggestion
    x = np.sort(x, axis=axis)
```